### PR TITLE
Full Rewrite - Alpha release

### PR DESCRIPTION
**Pull request by Bionded**
Full code rewrite
Added function to import zspotify api in your own projects
Add new arguments variables
Add argparse for parsing args
Fixes episode downloads
Fixes temp files (less storage io)
Now use only mutagen for setting tags for audio
Deleted archive for song, revert it in beta version

### CODE REVIEW

**Comment by nicknitewolf** [comment]
NoneType error when downloading playlists by url

**Comment by Bionded** [comment]
Can you write you launch arguments?

**Comment by Bionded** [comment]
Fixed

**Comment by nicknitewolf** [comment]
Directory arguments do not seem to be working

python main.py -d "~/Music/Spotify" -cd "~/Music/Spotify" -pl playlist_id
C:\Users\Nicholas/Music/ZSpotify Music/...

**Comment by nicknitewolf** [comment]
Migration from .song_archive to archive.json will be appreciated

File: main.py [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+        self.args = self.parse_args()
+        self.zs_api = zspotify_api(
+            sanitize=self.SANITIZE_CHARS,
+            config_dir=self.args.config_dir,
+            music_format=self.args.audio_format,
+            force_premium=self.args.force_premium,
+            anti_ban_wait_time=self.args.antiban_time,
+            credentials=self.args.credentials_file)
+        self.archive_file = os.path.join(self.args.config_dir, "archive.json")
+        self.download_dir = self.args.download_dir
```
Comment (nicknitewolf) [inline main.py:36]: This seems deprecated and not used anywhere?

Comment (Bionded) [reply main.py]: i want add new archives in next release.

**Comment by Bionded** [comment]
Fix download dir

**Comment by Bionded** [comment]
And yes windows can't undestand "~/"

**Comment by nicknitewolf** [comment]
I have hit another NoneType error with the same playlist. I am scared I will rate limit my account, could you take a look?

**Comment by Bionded** [comment]
> I have hit another NoneType error with the same playlist. I am scared I will rate limit my account, could you take a look?

checking

**Comment by Bionded** [comment]
can you send last 4-5 log lines?

**Comment by nicknitewolf** [comment]
> can you send last 4-5 log lines?

sorry not NoneType error, this is possibly caused by the download bar change

```
Traceback (most recent call last):
  File "C:\Users\Nicholas\Music\zspotify2\main.py", line 579, in <module>
    zs.start()
  File "C:\Users\Nicholas\Music\zspotify2\main.py", line 528, in start
    self.download_by_url(playlist)
  File "C:\Users\Nicholas\Music\zspotify2\main.py", line 358, in download_by_url
    ret = self.download_playlist(parsed_url['playlist'])
  File "C:\Users\Nicholas\Music\zspotify2\main.py", line 291, in download_playlist
    self.download_track(song['id'], basepath, "playlist")
  File "C:\Users\Nicholas\Music\zspotify2\main.py", line 262, in download_track
    bar.update(self.zs_api.progress['total'] - bar.n)
TypeError: 'bool' object is not subscriptable
```

**Comment by Bionded** [comment]
> I have hit another NoneType error with the same playlist. I am scared I will rate limit my account, could you take a look?
> 
> Edit: Error seems to come from songs added to playlist where the artist has deleted / hidden the song in Spotify. Hence all values present including user and date added except track and track_id

fix of this error i add in this commit
https://github.com/jsavargas/zspotify/commit/b7193df0e13948a4abfd54b0b7602ab29cc7ba1e

**Comment by Bionded** [comment]
and this error i fixed here https://github.com/jsavargas/zspotify/commit/b3a09031c5de08cc2dccd5a232ce0a7e38b6bd26
but commit message is bad, sorry(

**Comment by MauritzFunke** [comment]
@nicknitewolf Do you have a Linux machine to test too?

**Comment by nicknitewolf** [comment]
> @nicknitewolf Do you have a Linux machine to test too?

Yes I have WSL and another machine to test if needed. I am testing Windows flow only currently

**Comment by MauritzFunke** [comment]
If you need more machines to test, hit me up, I have 2 vms idleing around and one spare premium account.

**Comment by jsavargas** [comment]
Thank you very much, we could add a new branch, a new tag in docker and do the corresponding tests
To make sure everything works correctly before merging

**Comment by nicknitewolf** [comment]
> Thank you very much, we could add a new branch, a new tag in docker and do the corresponding tests To make sure everything works correctly before merging

This whole seems to drop song_archive support. Waiting for @Bionded to implement json version. It requires more work if we want to support migration from old CSV to json structure.

**Comment by Bionded** [comment]
Ok, Maybe need to create new branch where i can commit to branch in you repo. When implementation and testing done need only merge to master?
i need a one-two days to complete these things
1) archive and migration
2) search by podcasts and shows
3) add select for starting download song in playlist

**Comment by jsavargas** [comment]
I have created a branch from master called Bionded, to be able to test your changes there and to be able to generate a docker image later and test that everything works correctly

They are very significant changes.

**Comment by jsavargas** [comment]
I have generated a new docker image, based on your branch, calling it bionded-alfa (jsavargas/zspotify:bionded-alfa), in case anyone wants to try it directly in docker, just changing the tag in the compose should work without problems

jsavargas/zspotify:bionded-alfa

**Comment by jsavargas** [comment]
image

**Comment by Bionded** [comment]
I cant push in you branch ( No permissions). I will continue write in my repo. When i Done you need only copy my files to you. or megre that request

**Comment by Bionded** [comment]
And i realy don't understand why you need docker here, and can't test that flow. So, please, if someone can test and rewrite docker file - You are welcome)

**Comment by Bionded** [comment]
Archive added, with migration, need testing

**Comment by Bionded** [comment]
Little fix
And be careful, do a backup of old archive files. After migration complete old archive file being deleted.

**Comment by Bionded** [comment]
found that pull request
https://github.com/Bionded/zspotify/pull/2
@nicknitewolf Sorry for not answer. But it's realy good idea.

**Comment by Bionded** [comment]
if someone can, please, send me little part of your old archive, i will test too.

**Comment by nicknitewolf** [comment]
> if someone can, please, send me little part of your old archive, i will test too.

```
Found old archive, migrating to new one...
###   get_song_info - FAILED TO QUERY METADATA   ###
'tracks'
6BGbt6NGG5LGGqKdwAzzS8 {'error': {'status': 401, 'message': 'Invalid access token'}}
Traceback (most recent call last):
  File "C:\Users\Nicholas\Music\zspotify2\main.py", line 696, in <module>
    zs.start()
  File "C:\Users\Nicholas\Music\zspotify2\main.py", line 627, in start
    self.archive_migration()
  File "C:\Users\Nicholas\Music\zspotify2\main.py", line 282, in archive_migration
    musicinfo = self.zs_api.get_episode_info(id)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Nicholas\Music\zspotify2\zspotify.py", line 460, in get_episode_info
    for sum_px in info['images']:
                  ~~~~^^^^^^^^^^
KeyError: 'images'
```

Credentials not loaded before migration

**Comment by MauritzFunke** [comment]
@Bionded 
https://github.com/MauritzFunke/zspotify_archive
Uploaded a file with about 1700 entries to test

**Comment by Bionded** [comment]
> Credentials not loaded before migration

My bad( 
fixing

**Comment by Bionded** [comment]
> @Bionded https://github.com/MauritzFunke/zspotify_archive Uploaded a file with about 1700 entries to test

Thanks

**Comment by Bionded** [comment]
Dont use migration now, i rewrite it

**Comment by Bionded** [comment]
Rewrited and tested on this file https://github.com/MauritzFunke/zspotify_archive
Works fine, test on your archive

**Comment by MauritzFunke** [comment]
I just tried the new version. It's currently downloading already downloaded songs that are also in my old archive and also show up in the new archive.
Also the download bar is only filling to ~80 percent
```root@storage:/opt/python/zspotify_new# python3 main.py -ap

    ███████ ███████ ██████   ██████  ████████ ██ ███████ ██    ██
       ███  ██      ██   ██ ██    ██    ██    ██ ██       ██  ██
      ███   ███████ ██████  ██    ██    ██    ██ █████     ████
     ███         ██ ██      ██    ██    ██    ██ ██         ██
    ███████ ███████ ██       ██████     ██    ██ ██         ██
        
version: 2.0.0_alpha
Migration complete
Downloading Deep playlist
$uicideboy$ - Kill Yourself (Part III).mp3:  78%|██████████████████████████████████████████████████████████████████████████████████████████████████▊                            | 4.00M/5.14M [00:00<00:00, 14.0MB/s]
Converting $uicideboy$ - Kill Yourself (Part III).mp3
Added to archive: $uicideboy$ - Kill Yourself (Part III)
Set audiotags $uicideboy$ - Kill Yourself (Part III).mp3
Finished downloading $uicideboy$ - Kill Yourself (Part III).mp3
$uicideboy$ - Paris.mp3:  80%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████                             | 3.75M/4.68M [00:00<00:00, 13.1MB/s]
Converting $uicideboy$ - Paris.mp3
Added to archive: $uicideboy$ - Paris
Set audiotags $uicideboy$ - Paris.mp3
Finished downloading $uicideboy$ - Paris.mp3
$uicideboy$ - 2nd Hand.mp3:  90%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎             | 5.12M/5.67M [00:00<00:00, 17.9MB/s]
Converting $uicideboy$ - 2nd Hand.mp3
Added to archive: $uicideboy$ - 2nd Hand
Set audiotags $uicideboy$ - 2nd Hand.mp3
Finished downloading $uicideboy$ - 2nd Hand.mp3
$uicideboy$ - Either Hated or Ignored.mp3:  82%|████████████████████████████████████████████████████████████████████████████████████████████████████████▌                       | 4.37M/5.36M [00:00<00:00, 15.3MB/s]
Converting $uicideboy$ - Either Hated or Ignored.mp3
Added to archive: $uicideboy$ - Either Hated or Ignored
Set audiotags $uicideboy$ - Either Hated or Ignored.mp3
Finished downloading $uicideboy$ - Either Hated or Ignored.mp3
$uicideboy$ - Mount Sinai.mp3:  68%|██████████████████████████████████████████████████████████████████████████████████████████████▋                                             | 3.62M/5.36M [00:00<00:00, 12.6MB/s]
Converting $uicideboy$ - Mount Sinai.mp3
Added to archive: $uicideboy$ - Mount Sinai
Set audiotags $uicideboy$ - Mount Sinai.mp3
Finished downloading $uicideboy$ - Mount Sinai.mp3
$uicideboy$ - Magazine.mp3:  70%|███████████████████████████████████████████████████████████████████████████████████████████████████▊                                           | 3.62M/5.19M [00:00<00:00, 12.6MB/s]
Converting $uicideboy$ - Magazine.mp3
Added to archive: $uicideboy$ - Magazine
Set audiotags $uicideboy$ - Magazine.mp3
Finished downloading $uicideboy$ - Magazine.mp3
$uicideboy$ - O Pana!.mp3:  81%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▍                           | 4.25M/5.25M [00:00<00:00, 11.1MB/s]
Converting $uicideboy$ - O Pana!.mp3
Added to archive: $uicideboy$ - O Pana!
Set audiotags $uicideboy$ - O Pana!.mp3
Finished downloading $uicideboy$ - O Pana!.mp3
$uicideboy$ - Lte.mp3:  72%|██████████████████████████████████████████████████████████████████████████████████████████████████████████▊                                         | 4.00M/5.54M [00:00<00:00, 20.9MB/s]
Converting $uicideboy$ - Lte.mp3
Added to archive: $uicideboy$ - Lte
Set audiotags $uicideboy$ - Lte.mp3
Finished downloading $uicideboy$ - Lte.mp3
$uicideboy$ - Memoirs of a Gorilla.mp3:  65%|█████████████████████████████████████████████████████████████████████████████████████                                              | 3.25M/5.00M [00:00<00:00, 17.0MB/s]
Converting $uicideboy$ - Memoirs of a Gorilla.mp3
Added to archive: $uicideboy$ - Memoirs of a Gorilla
Set audiotags $uicideboy$ - Memoirs of a Gorilla.mp3
Finished downloading $uicideboy$ - Memoirs of a Gorilla.mp3
Night Lovell, $uicideboy$ - JOAN OF ARC.mp3:  71%|████████████████████████████████████████████████████████████████████████████████████████▉                                     | 5.00M/7.08M [00:00<00:00, 17.4MB/s]
Converting Night Lovell, $uicideboy$ - JOAN OF ARC.mp3
Added to archive: Night Lovell, $uicideboy$ - JOAN OF ARC
Set audiotags Night Lovell, $uicideboy$ - JOAN OF ARC.mp3
Finished downloading Night Lovell, $uicideboy$ - JOAN OF ARC.mp3
$uicideboy$ - Fuckthepopulation.mp3:  97%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉    | 6.37M/6.58M [00:00<00:00, 16.7MB/s]
Converting $uicideboy$ - Fuckthepopulation.mp3
Added to archive: $uicideboy$ - Fuckthepopulation
Set audiotags $uicideboy$ - Fuckthepopulation.mp3
Finished downloading $uicideboy$ - Fuckthepopulation.mp3
$uicideboy$ - Diemonds.mp3:  80%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋                             | 4.75M/5.97M [00:00<00:00, 16.6MB/s]
Converting $uicideboy$ - Diemonds.mp3
Added to archive: $uicideboy$ - Diemonds
Set audiotags $uicideboy$ - Diemonds.mp3
Finished downloading $uicideboy$ - Diemonds.mp3
$uicideboy$ - Champion of Death.mp3:  92%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎          | 4.75M/5.16M [00:00<00:00, 16.6MB/s]
Converting $uicideboy$ - Champion of Death.mp3
Added to archive: $uicideboy$ - Champion of Death
Set audiotags $uicideboy$ - Champion of Death.mp3
Finished downloading $uicideboy$ - Champion of Death.mp3
$uicideboy$ - My Flaws Burn Through My Skin Like Demonic Flames from Hell.mp3:  90%|██████████████████████████████████████████████████████████████████████████████████▋         | 5.75M/6.40M [00:00<00:00, 15.0MB/s]
Converting $uicideboy$ - My Flaws Burn Through My Skin Like Demonic Flames from Hell.mp3
Added to archive: $uicideboy$ - My Flaws Burn Through My Skin Like Demonic Flames from Hell
Set audiotags $uicideboy$ - My Flaws Burn Through My Skin Like Demonic Flames from Hell.mp3
Finished downloading $uicideboy$ - My Flaws Burn Through My Skin Like Demonic Flames from Hell.mp3
$uicideboy$ - Do You Believe in God.mp3:  71%|████████████████████████████████████████████████████████████████████████████████████████████                                      | 3.25M/4.59M [00:00<00:00, 17.0MB/s]
Converting $uicideboy$ - Do You Believe in God.mp3
Added to archive: $uicideboy$ - Do You Believe in God
Set audiotags $uicideboy$ - Do You Believe in God.mp3
Finished downloading $uicideboy$ - Do You Believe in God.mp3
$uicideboy$ - Youre Now Tuning into 66.6 FM with DJ Rapture (The Hottest Hour of the Evening).mp3:  90%|████████████████████████████████████████████████████████████████▋       | 5.75M/6.40M [00:00<00:00, 20.0MB/s]
Converting $uicideboy$ - Youre Now Tuning into 66.6 FM with DJ Rapture (The Hottest Hour of the Evening).mp3
Added to archive: $uicideboy$ - Youre Now Tuning into 66.6 FM with DJ Rapture (The Hottest Hour of the Evening)
Set audiotags $uicideboy$ - Youre Now Tuning into 66.6 FM with DJ Rapture (The Hottest Hour of the Evening).mp3
Finished downloading $uicideboy$ - Youre Now Tuning into 66.6 FM with DJ Rapture (The Hottest Hour of the Evening).mp3
$uicideboy$, Pouya - $Outh $Ide $Uicide.mp3:  72%|███████████████████████████████████████████████████████████████████████████████████████████▏                                  | 5.12M/7.08M [00:00<00:00, 17.9MB/s]
Converting $uicideboy$, Pouya - $Outh $Ide $Uicide.mp3
Added to archive: $uicideboy$, Pouya - $Outh $Ide $Uicide
^CInterrupted by user
```

**Comment by MauritzFunke** [comment]
For example the song "$uicideboy$ - Paris.mp3"
```
Finished downloading $uicideboy$ - Paris.mp3
$uicideboy$ - 2nd Hand.mp3:  90%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎             | 5.12M/5.67M [00:00<00:00, 17.9MB/s]
```
`ls -l `:
`-rw-r--r-- 1 root root 4558414 25. Feb 14:52 '$uicideboy$ - Paris.mp3'`
`ls -lh`:
`-rw-r--r-- 1 root root 4,4M 25. Feb 14:52 '$uicideboy$ - Paris.mp3'`

**Comment by Bionded** [comment]
for ignoring file has already in arhive need set -ss flag

**Comment by MauritzFunke** [comment]
Thats sadly only working for files downloaded after the migrations. Probably because the "fullpath" is not set for the old songs. Maybe add a option to search for the filename and add that to the new archive as well?

**Comment by nicknitewolf** [comment]
> Thats sadly only working for files downloaded after the migrations. Probably because the "fullpath" is not set for the old songs. Maybe add a option to search for the filename and add that to the new archive as well?

Could you send the migrated json? skip files depends on track_id rather than fullpath so I am not so sure why it is not working for you

**Comment by MauritzFunke** [comment]
So I just compared the ids. That were saved.
```
6tO5bxNnMuh1c3cziSpecA
4irYeuAi87yyGHcI4h9s0x
```

I have no quick way of accessing the API so I have to guess. Maybe the one is the single version and the other the album version?

I'll download a lot of playlists again to test if its working correctly

**Comment by nicknitewolf** [comment]
> So I just compared the ids. That were saved.
> 
> ```
> 6tO5bxNnMuh1c3cziSpecA
> 4irYeuAi87yyGHcI4h9s0x
> ```
> 
> I have no quick way of accessing the API so I have to guess. Maybe the one is the single version and the other the album version?
> 
> I'll download a lot of playlists again to test if its working correctly

image
image

Yes you are partially right, one is explicit the other isnt

**Comment by nicknitewolf** [comment]
<h3>Potential feature improvements</h3>
<table>
<tr><td>
Album file naming convention is a lot different from playlist naming convention (same for previous versions). Seems inconsistent.
Downloading an album with a few songs previously downloaded makes file structure very inconsistent. Should the file be renamed and placed back to album folder?

For example:

Playlist A
Album B Artist - Song 3.mp3

Album B
1. Song 1.mp3
2. Song 2.mp3
4. Song 4.mp3
</td></tr>

<tr><td>Save argparse userflags to config file</td></tr>

<tr><td>Make sure download bar reaches 100%</td></tr>
<tr><td>Use ID3v2.3 tagging on mutagen for better compatibility with bad players (windows >:( player)</td></tr>
<tr><td>Comment actual track url instead of weird id[spotify.com:track:6tO5bxNnMuh1c3cziSpecA]</td></tr>
</table>

**Comment by MauritzFunke** [comment]
Please be sure to migrate the track ids. I'd be happy to help. Are there other communication channels to exchange?

File: main.py [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+            tags["TDOR"] = id3.TDOR(encoding=3, text=release_year)  # TDOR Original release time
+        if disc_number is not None:
+            tags["TPOS"] = id3.TPOS(encoding=3, text=str(disc_number))  # TPOS Part of a set
+        if track_number is not None:
+            tags["TRCK"] = id3.TRCK(
+                encoding=3, text=str(track_number)
+            )  # TRCK Track number/Position in set
+        if track_id_str is not None:
+            tags["COMM"] = id3.COMM(
+                encoding=3, lang="eng", text="id[spotify.com:track:" + track_id_str + "]"
```
Comment (nicknitewolf) [inline main.py:275]: encoding=3, lang="eng", text="https://open.spotify.com/track/spotify.com/" + track_id_str

Comment (Bionded) [reply main.py:275]: Ok, change that

**Comment by Bionded** [comment]
fix -s flag (now is -sn, -ss is now -s
possible 100% download bar fix
add spotify url in track tags

**Comment by Bionded** [comment]
i can do a manual format for audio naming.
 somthing like that in conf file.
 playlist_audio = {artist_name} - {song_name}.{ext}
 album_audio = {track_num}.  {song_name}.{ext}
 But not in that release, please. Need release beta, fix all bugs, and after add new features.

**Comment by Bionded** [comment]
1) Find files on migration archive, i think be a realy bugged. I don't like that functionality. If need - ok, i can write different script only for you, but don't wanna add this in main functionality.
2) argparse to config file - good idea, but too, not in this release.

**Comment by Bionded** [comment]
> Please be sure to migrate the track ids

Track ids now migrating to new archive. Something wrong with that?

**Comment by jsavargas** [comment]
I will bring these changes to a new branch, and there we will improve together, when everything is debugged, we take master, do you think?

This is how I correct the dockerfile and other things, I'm interested that it also works without parameters, so as not to affect the functionalities of before

**Comment by jsavargas** [comment]
Added to this branch, corrected the dockerfile and I would like to review after some options have the alternative in environment variables, in order to preserve the functionality in docker, since that is where I use it the most and the people I know, you can look for me on telegram Or we could make a discord of this project, together with all the people who have been collaborating over time

https://github.com/jsavargas/zspotify/tree/Bionded


image

**Comment by jsavargas** [comment]
image

The download was at 99%, I don't know if it finished completely or it was missing, I will download it with the previous version and compare, something is not quite right

**Comment by nicknitewolf** [comment]
> image
> 
> The download was at 99%, I don't know if it finished completely or it was missing, I will download it with the previous version and compare, something is not quite right

I think it’s a race condition where the file is already downloaded and has moved on to the converting step before the bar updates. The songs I have downloaded still have their full lengths intact

**Comment by jsavargas** [comment]
ah, it may be, I had corrected that in the version that is in master, I will compare the methods in case it can be improved

**Comment by Bionded** [comment]
Add that
https://github.com/jsavargas/zspotify/pull/75

**Pull request merged**